codegen/files/rest/server.go

```diff
@@ -1,11 +1,587 @@
 package rest
 
 import (
+	"fmt"
+	"text/template"
+
 	"goa.design/goa.v2/codegen"
+	"goa.design/goa.v2/codegen/files"
+	"goa.design/goa.v2/design"
 	"goa.design/goa.v2/design/rest"
+	"net/http"
+	"sort"
+	"strings"
+)
+
+//todo: in encode headers and mappings
+
+const serverStructT = `{{ printf "%s lists the %s service endpoint HTTP handlers." .HandlersStruct .ServiceName | comment }}
+type {{ .HandlersStruct }} struct {
+	{{- range .ActionData }}
+	{{ .EndpointName }} http.Handler
+	{{- end }}
+}
+`
+
+const serverConstructorT = `{{ printf "%s instantiates HTTP handlers for all the %s service endpoints." .Constructor .ServiceName | comment }}
+func {{ .Constructor }}(
+	e *endpoints.{{ .ServiceName }},
+	dec rest.RequestDecoderFunc,
+	enc rest.ResponseEncoderFunc,
+	logger goa.Logger,
+) *{{ .HandlersStruct }} {
+	return &{{ .HandlersStruct }}{
+		{{- range .ActionData }}
+		{{ .EndpointName }}: {{ .Constructor }}(e.{{ .EndpointName }}, dec, enc, logger),
+		{{- end }}
+	}
+}
+`
+
+const serverMountT = `{{ printf "%s configures the mux to serve the %s endpoints." .MountHandlers .ServiceName | comment }}
+func {{ .MountHandlers }}(mux rest.ServeMux, h *{{ .HandlersStruct }}) {
+	{{- range .ActionData }}
+	{{ .MountHandler }}(mux, h.{{ .EndpointName }})
+	{{- end }}
+}
+`
+
+const serverHandlerT = `{{ printf "%s configures the mux to serve the \"%s\" service \"%s\" endpoint." .MountHandler .ServiceName .EndpointName | comment }}
+func {{ .MountHandler }}(mux rest.ServeMux, h http.Handler) {
+	{{- range .Routes }}
+	mux.Handle("{{ .Method }}", "{{ .Path }}", h)
+	{{- end }}
+}
+`
+
+const serverHandlerConstructorT = `{{ printf "%s creates a HTTP handler which loads the HTTP request and calls the \"%s\" service \"%s\" endpoint." .Constructor .ServiceName .EndpointName | comment }}
+{{ comment "The middleware is mounted so it executes after the request is loaded and thus may access the request state via the rest package ContextXXX functions."}}
+func {{ .Constructor }}(
+	endpoint goa.Endpoint,
+	dec rest.RequestDecoderFunc,
+	enc rest.ResponseEncoderFunc,
+	logger goa.Logger,
+) http.Handler {
+	var (
+		decodeRequest  = {{ .Decoder }}(dec)
+		encodeResponse = {{ .Encoder }}EncodeResponse(enc)
+		encodeError    = {{ if .HasErrors }}{{ .ErrorEncoder }}{{ else }}EncodeError{{ end }}(enc, logger)
+	)
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		payload, err := decodeRequest(r)
+		if err != nil {
+			encodeError(w, r, goa.ErrInvalid("request invalid: %s", err))
+			return
+		}
+
+		res, err := endpoint(r.Context(), payload)
+
+		if err != nil {
+			encodeError(w, r, err)
+			return
+		}
+		if err := encodeResponse(w, r, res); err != nil {
```

Comment:
So this touches on something that is quite fundamental and that is whether goa should enforce that the attributes used to define the HTTP endpoints params, bodies and headers must be a subset of the (non-http) endpoint payload and result types.

That's why I wrote *endpoint* result and not HTTP response. If we go with "all attributes used to define HTTP elements must be defined in the corresponding endpoint type (payload for HTTP request elements, result for HTTP response elements) then simply checking that the endpoint result type is `design.Empty` is enough to know that the response body is empty. Otherwise need to check whether all the attributes of the endpoint result are used to define response headers and if not the remaining attributes define the response body. See the DSL spec for examples.

Another approach would consist of making it possible to define attributes in the HTTP DSL that is not defined in the Endpoint DSL. There is a major drawback to this approach though which is that it makes impossible to generate the code that maps the service interface types to the HTTP types. The generated code may not be able to completely form the HTTP response because it's missing the extra fields that were HTTP specific from the service result type. The user would have to modify the generated type to initialize these values. Stepping back a little this breaks the nice layer abstraction.

So at this point my preference would be for the DSL engine to enforce that attributes used to define HTTP elements exist in the corresponding non-HTTP type. There may be extra validations / examples defined in the HTTP specific attributes but the names must match attributes from the payload (request) or result (response) type.